dex/networks/ltc/block.go

```diff
@@ -0,0 +1,92 @@
+// also available online at https://blueoakcouncil.org
+
+package ltc
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	"github.com/btcsuite/btcd/wire"
+)
+
+const (
+	// mwebVer is the bit of the block header's version that indicates the
+	// presence of a MWEB.
+	mwebVer = 0x20000000 // 1 << 29
+
+	// EmptyTxHash is the txid of an empty wire.MsgTx that may be included in a
+	// block with a MW integration transaction (HogEx).
+	EmptyTxHash = "f702453dd03b0f055e5437d76128141803984fb10acb85fc3b2184fae2f3fa78"
+)
+
+// DeserializeBlock decodes the bytes of a serialized Litecoin block. This
+// function exists because MWEB changes both the block and transaction
+// serializations. Blocks may have a MW "extension block" for "peg-out"
+// transactions, and this EB is after all the transactions in the regular LTC
+// block. After the canonical transactions in the regular block, there may be
+// zero or more "peg-in" transactions followed by one integration transactions
+// (also known as a HogEx transaction), all still in the regular LTC block. The
+// peg-in txns decode correctly, but the integration tx is a special transaction
+// with the witness tx flag with bit 3 set (8), which prevents correct
+// wire.MsgTx deserialization. These integration transactions are replaced with
+// a dummy MsgTx recognizable by EmptyTxHash. We make no effort to decode the EW
+// or any of the peg-out transactions, both of which have unique encodings.
+// Refs:
+// https://github.com/litecoin-project/lips/blob/master/lip-0002.mediawiki#PegOut_Transactions
+// https://github.com/litecoin-project/lips/blob/master/lip-0003.mediawiki#Specification
+// https://github.com/litecoin-project/litecoin/commit/9d1f530a5fa6d16871fdcc3b506be42b593d3ce4
+// https://github.com/litecoin-project/litecoin/commit/8c82032f45e644f413ec5c91e121a31c993aa831
+// (src/libmw/include/mw/models/tx/Transaction.h for the `mweb_tx` field of the
+// `CTransaction` in the "primitives" commit).
+func DeserializeBlock(blk io.Reader) (*wire.MsgBlock, error) {
+	// Block header
+	hdr := &wire.BlockHeader{}
+	err := hdr.Deserialize(blk)
+	if err != nil {
+		return nil, fmt.Errorf("failed to deserialize block header: %w", err)
+	}
+
+	// May contain a MWEB *after* the txns.
+	containsMWEB := hdr.Version&mwebVer != 0
+
+	// This block's transactions
+	txnCount, err := wire.ReadVarInt(blk, 0)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse transaction count: %w", err)
+	}
+
+	// We can only decode the canonical txns, not the integ/hogex txn at the end
+	// or the mw peg-in txs in the EB.
+	txns := make([]*wire.MsgTx, 0, int(txnCount))
+	for i := 0; i < cap(txns); i++ {
+		msgTx := &wire.MsgTx{}
+		err = msgTx.Deserialize(blk)
+		if err != nil {
+			if !containsMWEB {
+				return nil, fmt.Errorf("failed to deserialize transaction %d of %d in block %v: %w",
+					i+1, txnCount, hdr.BlockHash(), err)
+			}
+			break // the last tx should be an integration tx (hogex)
```

Comment:
Worth a trace log? Deserialize could possible fail for a different reason that what we expect.